File: crypto/src/main/java/io/oversec/one/crypto/symbase/KeyCache.java
package io.oversec.one.crypto.symbase;

import android.app.AlarmManager;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import io.oversec.one.crypto.sym.KeyNotCachedException;
import io.oversec.one.crypto.sym.KeystoreIntentService;
import io.oversec.one.crypto.sym.SymmetricKeyPlain;
import io.oversec.one.crypto.sym.ui.UnlockKeyActivity;
import roboguice.util.Ln;

import java.io.IOException;
import java.util.*;

public class KeyCache {
    private static final String EXTRA_KEY_ID = "EXTRA_KEY_ID";
    private static final String OVERSEC_BROADCAST_ACTION_EXPIRE_KEY = "OVERSEC_BROADCAST_ACTION_EXPIRE_KEY";
    private static KeyCache INSTANCE;

    private final AlarmManager mAlarmManager;
    private final NotificationManager mNotificationManager;
    private final List<OversecKeyCacheListener> mKeyCacheListeners = new ArrayList<>();
    private final Map<Long, PendingIntent> mPendingAlarms = new HashMap<>();
    private final Map<Long, SymmetricKeyPlain> mKeyMap = new HashMap<>();
    private final Set<Long> mExpireOnScreenOff = new HashSet<>();
    private final Context mCtx;

    private int requestCount = 0;

    public static synchronized KeyCache getInstance(Context ctx) {
        if (INSTANCE == null) {
            INSTANCE = new KeyCache(ctx);
        }
        return INSTANCE;
    }

    private KeyCache(Context ctx) {
        mCtx = ctx;
        mAlarmManager = (AlarmManager) mCtx.getSystemService(Context.ALARM_SERVICE);
        mNotificationManager = (NotificationManager) mCtx.getSystemService(Context.NOTIFICATION_SERVICE);
        BroadcastReceiver aIntentReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();

                if (action.equals(OVERSEC_BROADCAST_ACTION_EXPIRE_KEY)) {
                    long keyId = intent.getLongExtra(EXTRA_KEY_ID, -1);
                    expire(keyId);
                }

                if (action.equals(Intent.ACTION_SCREEN_OFF)) {
                    expireOnScreenOff();
                }
            }
        };

        IntentFilter filter = new IntentFilter();
        filter.addAction(OVERSEC_BROADCAST_ACTION_EXPIRE_KEY);
        filter.addAction(Intent.ACTION_SCREEN_OFF);
        mCtx.registerReceiver(aIntentReceiver, filter);
    }

    public synchronized void doCacheKey(SymmetricKeyPlain plainKey, long ttl) {
        long keyId = plainKey.getId();


        if (mKeyMap.containsKey(keyId)) {
            Ln.d("already cached!");
            //ToDo: this should never happen, but if so, maybe we should update the TTL ?

        } else {

            mKeyMap.put(keyId, plainKey);


            for (OversecKeyCacheListener l : mKeyCacheListeners) {
                l.onStartedCachingKey(keyId);
            }

            mNotificationManager.notify(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS,
                    KeystoreIntentService.buildCachedKeysNotification(mCtx, getAllCachedKeyAliases()));

            if (ttl == 0) {
                mExpireOnScreenOff.add(keyId);
            } else if (ttl < Integer.MAX_VALUE) {

                Intent intent = new Intent(OVERSEC_BROADCAST_ACTION_EXPIRE_KEY);
                intent.putExtra(EXTRA_KEY_ID, keyId);
                // request code should be unique for each PendingIntent, thus keyId is used
                PendingIntent alarmIntent = PendingIntent.getBroadcast(mCtx, ++requestCount, intent,
                        PendingIntent.FLAG_CANCEL_CURRENT);
                mPendingAlarms.put(keyId, alarmIntent);

                long triggerTime = new Date().getTime() + ttl * 1000;
                mAlarmManager.set(AlarmManager.RTC_WAKEUP, triggerTime, alarmIntent);
            }
        }


    }

    private List<String> getAllCachedKeyAliases() {
        ArrayList<String> res = new ArrayList();
        for (SymmetricKeyPlain k : mKeyMap.values()) {
            res.add(k.getName());
        }
        Collections.sort(res);
        return res;
    }

    public synchronized void clearAll() {
        Set<Long> ids = new HashSet<>(mKeyMap.keySet());
        for (Long id : ids) {
            expire(id);
        }
        mKeyMap.clear();
        mNotificationManager.cancel(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS);
    }

    public synchronized void expire(long keyId) {
        PendingIntent pi = mPendingAlarms.get(keyId);
        if (pi != null) {
            mAlarmManager.cancel(pi);
        }
        mExpireOnScreenOff.remove(keyId);
        SymmetricKeyPlain k = mKeyMap.get(keyId);
        if (k != null) {
            //TODO: why does this happen? Maybe restarting the app after a crash and alarms still being in place?
            k.clearKeyData();
            mKeyMap.remove(keyId);
            for (OversecKeyCacheListener l : mKeyCacheListeners) {
                l.onFinishedCachingKey(keyId);
            }

        }


        if (mKeyMap.size() > 0) {
            mNotificationManager.notify(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS,
                    KeystoreIntentService.buildCachedKeysNotification(mCtx, getAllCachedKeyAliases()));
        } else {
            mNotificationManager.cancel(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS);
        }
    }

    synchronized void expireOnScreenOff() {
        Set<Long> ids = new HashSet<>(mExpireOnScreenOff);
        for (Long id : ids) {
            expire(id);
        }
        mExpireOnScreenOff.clear();
    }

    public synchronized SymmetricKeyPlain get(Long keyId) throws KeyNotCachedException {
        SymmetricKeyPlain res = mKeyMap.get(keyId);
        if (res == null) {
            throw new KeyNotCachedException(UnlockKeyActivity.buildPendingIntent(mCtx, keyId));
        }
        return res;
    }

    public boolean hasKey(Long keyId) {
        return mKeyMap.containsKey(keyId);
    }

    public synchronized void addKeyCacheListener(OversecKeyCacheListener l) {
        mKeyCacheListeners.add(l);
    }

    public synchronized void removeKeyCacheListener(OversecKeyCacheListener l) {
        mKeyCacheListeners.remove(l);

    }


    public synchronized SymmetricKeyPlain getKeyByHashedKeyId(long keyhash, byte[] salt, int costKeyhash) {

        Set<Long> allIds = mKeyMap.keySet();
        for (Long id : allIds) {
            long aSessionKeyId = KeyUtil.calcSessionKeyId(id, salt, costKeyhash);

            if (aSessionKeyId == keyhash) {
                return mKeyMap.get(id);
            }
        }

        return null;
    }


    public List<SymmetricKeyPlain> getAllSimpleKeys() {
        ArrayList<SymmetricKeyPlain> res = new ArrayList<>();
        for (SymmetricKeyPlain k : mKeyMap.values()) {
            if (k.isSimpleKey()) {
                res.add(k);
            }
        }
        Collections.sort(res, new Comparator<SymmetricKeyPlain>() {
            @Override
            public int compare(SymmetricKeyPlain lhs, SymmetricKeyPlain rhs) {
                return lhs.getCreatedDate().compareTo(rhs.getCreatedDate());
            }
        });
        return res;
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/symbase/KeyCache.kt
package io.oversec.one.crypto.symbase

import android.annotation.SuppressLint
import android.app.AlarmManager
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import io.oversec.one.crypto.sym.KeyNotCachedException
import io.oversec.one.crypto.sym.KeystoreIntentService
import io.oversec.one.crypto.sym.SymmetricKeyPlain
import io.oversec.one.crypto.sym.ui.UnlockKeyActivity
import roboguice.util.Ln
import java.util.*

class KeyCache private constructor(private val mCtx: Context) {

    private val mAlarmManager = mCtx.getSystemService(Context.ALARM_SERVICE) as AlarmManager
    private val mNotificationManager =
        mCtx.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    private val mKeyCacheListeners = ArrayList<OversecKeyCacheListener>()
    private val mPendingAlarms = HashMap<Long, PendingIntent>()
    private val mKeyMap = HashMap<Long, SymmetricKeyPlain>()
    private val mExpireOnScreenOff = HashSet<Long>()

    private var requestCount = 0

    private val allCachedKeyAliases: List<String>
        get() {
            return mKeyMap.mapValues {
                it.value.name
            }.values.filterNotNull().sortedBy {
                it
            }

        }


    val allSimpleKeys: List<SymmetricKeyPlain>
        get() {
            return mKeyMap.filterValues {
                it.isSimpleKey
            }.values.sortedBy {
                it.createdDate
            }
        }

    init {
        val aIntentReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                val action = intent.action

                if (action == OVERSEC_BROADCAST_ACTION_EXPIRE_KEY) {
                    val keyId = intent.getLongExtra(EXTRA_KEY_ID, -1)
                    expire(keyId)
                }

                if (action == Intent.ACTION_SCREEN_OFF) {
                    expireOnScreenOff()
                }
            }
        }

        val filter = IntentFilter()
        filter.addAction(OVERSEC_BROADCAST_ACTION_EXPIRE_KEY)
        filter.addAction(Intent.ACTION_SCREEN_OFF)
        mCtx.registerReceiver(aIntentReceiver, filter)
    }

    @Synchronized
    fun doCacheKey(plainKey: SymmetricKeyPlain, ttl: Long) {
        val keyId = plainKey.id

        if (mKeyMap.containsKey(keyId)) {
            Ln.d("already cached!")
            //ToDo: this should never happen, but if so, maybe we should update the TTL ?

        } else {

            mKeyMap[keyId] = plainKey
            mKeyCacheListeners.forEach {
                it.onStartedCachingKey(keyId)
            }

            mNotificationManager.notify(
                KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS,
                KeystoreIntentService.buildCachedKeysNotification(mCtx, allCachedKeyAliases)
            )

            if (ttl == 0L) {
                mExpireOnScreenOff.add(keyId)
            } else if (ttl < Integer.MAX_VALUE) {

                val intent = Intent(OVERSEC_BROADCAST_ACTION_EXPIRE_KEY)
                intent.putExtra(EXTRA_KEY_ID, keyId)
                // request code should be unique for each PendingIntent, thus keyId is used
                val alarmIntent = PendingIntent.getBroadcast(
                    mCtx, ++requestCount, intent,
                    PendingIntent.FLAG_CANCEL_CURRENT
                )
                mPendingAlarms[keyId] = alarmIntent

                val triggerTime = Date().time + ttl * 1000
                mAlarmManager.set(AlarmManager.RTC_WAKEUP, triggerTime, alarmIntent)
            }
        }


    }

    @Synchronized
    fun clearAll() {
        mKeyMap.keys.forEach {
            expire(it)
        }
        mKeyMap.clear()
        mNotificationManager.cancel(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS)
    }

    @Synchronized
    fun expire(keyId: Long) {
        val pi = mPendingAlarms[keyId]
        if (pi != null) {
            mAlarmManager.cancel(pi)
        }
        mExpireOnScreenOff.remove(keyId)
        val k = mKeyMap[keyId]
        if (k != null) {
            //TODO: why does this happen? Maybe restarting the app after a crash and alarms still being in place?
            k.clearKeyData()
            mKeyMap.remove(keyId)
            for (l in mKeyCacheListeners) {
                l.onFinishedCachingKey(keyId)
            }
        }

        if (mKeyMap.size > 0) {
            mNotificationManager.notify(
                KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS,
                KeystoreIntentService.buildCachedKeysNotification(mCtx, allCachedKeyAliases)
            )
        } else {
            mNotificationManager.cancel(KeystoreIntentService.NOTIFICATION_ID__CACHED_KEYS)
        }
    }

    @Synchronized
    internal fun expireOnScreenOff() {
        val ids = HashSet(mExpireOnScreenOff)
        for (id in ids) {
            expire(id!!)
        }
        mExpireOnScreenOff.clear()
    }

    @Synchronized
    @Throws(KeyNotCachedException::class)
    operator fun get(keyId: Long?): SymmetricKeyPlain {
        return mKeyMap[keyId] ?: throw KeyNotCachedException(
            UnlockKeyActivity.buildPendingIntent(
                mCtx,
                keyId
            )
        )
    }

    fun hasKey(keyId: Long?): Boolean {
        return mKeyMap.containsKey(keyId)
    }

    @Synchronized
    fun addKeyCacheListener(l: OversecKeyCacheListener) {
        mKeyCacheListeners.add(l)
    }

    @Synchronized
    fun removeKeyCacheListener(l: OversecKeyCacheListener) {
        mKeyCacheListeners.remove(l)
    }


    @Synchronized
    fun getKeyByHashedKeyId(keyhash: Long, salt: ByteArray, costKeyhash: Int): SymmetricKeyPlain? {
        return mKeyMap.keys.find {
            val aSessionKeyId = KeyUtil.calcSessionKeyId(it, salt, costKeyhash)
            aSessionKeyId == keyhash
        }?.let {
            mKeyMap[it]
        }
    }

    companion object {
        private const val EXTRA_KEY_ID = "EXTRA_KEY_ID"
        private const val OVERSEC_BROADCAST_ACTION_EXPIRE_KEY =
            "OVERSEC_BROADCAST_ACTION_EXPIRE_KEY"


        @SuppressLint("StaticFieldLeak") // note that we're storing *Application*context
        @Volatile
        private var INSTANCE: KeyCache? = null

        fun getInstance(ctx: Context): KeyCache =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: KeyCache(ctx.applicationContext).also { INSTANCE = it }
            }
    }
}